server/conn.go

```diff
@@ -348,9 +353,29 @@ func (cc *clientConn) writeInitialHandshake(ctx context.Context) error {
 	data = append(data, cc.salt[8:]...)
 	data = append(data, 0)
 	// auth-plugin name
-	data = append(data, []byte(mysql.AuthNativePassword)...)
+	if cc.ctx == nil {
+		err := cc.openSession()
+		if err != nil {
+			return err
+		}
+	}
+	defAuthPlugin, err := variable.GetGlobalSystemVar(cc.ctx.GetSessionVars(), variable.DefaultAuthPlugin)
+	if err != nil {
+		return err
+	}
+	cc.authPlugin = defAuthPlugin
+	data = append(data, []byte(defAuthPlugin)...)
+
+	// Close the session to force this to be re-opened after we parse the response. This is needed
+	// to ensure we use the collation and client flags from the response for the session.
+	err = cc.ctx.Close()
+	if err != nil {
+		return err
+	}
+	cc.ctx = nil
```

Comment:
Just a hint: open a new session may need to load the global variables (maybe not true when they're cached), and do it once again could slow down the connection speed.

Reply:
I agree with your comment.

The cache is improved with sysvar cache, but it is still expensive. The current code will re-run validation on all of the sysvars. This can be improved though: in future we can cache the validated values in the sysvar cache, and when a session is initialized just copy the sysvar cache map to it.

The complicated case will be eliminating the need to call the 'init functions' (i.e. calling `sv.SetSession()` so that `s.VarXXX` is set.) I think we can remove a lot of `SetSession()` funcs by adding something like `sv.ToBool()`, `sv.ToUint64()`.

Reply:
Indeed, creating a new session for handshake is relatively heavy, indirectly leading to the occurrence of https://github.com/pingcap/tidb/issues/44502.